src/v/archival/archival_metadata_stm.h

```diff
@@ -94,6 +94,13 @@ class command_batch_builder {
       cloud_storage::scrub_status status,
       cloud_storage::anomalies detected);
     command_batch_builder& reset_scrubbing_metadata();
+    /// Add read-write fence
+    ///
+    /// The fence prevents all subsequent commands in the batch from being
+    /// applied to the in-memory state of the STM if the 'applied_offset' is
+    /// greater than the 'offset'. This mechanism is supposed to be used as a
+    /// concurrency-control mechanism.
+    command_batch_builder& read_write_fence(model::offset offset);
```

Comment:
Just curious, is the only reason to allow an input offset for testing? It may be less error prone to just call into the stm for the last applied offset, and have tests explicitly interleave construction of the builders (don't feel strongly about it though)

Reply:
It's not for testing. The STM doesn't know when you accessed the manifest so it can't start building the batch at the right moment. The user is supposed to read the fence offset manually and then use this method to add it to the batch.

Reply:
You're also allowed not to use it. For instance, in the unsafe manifest reset you should just avoid using it and let possible concurrent operation to be interrupted.

Reply:
Ah that makes sense. Thanks for explaining